tests/utils/pytest_parallel_gpu.py

```
@@ -445,6 +451,52 @@ def _cap(gi: int) -> float:
+    from tests.utils.managed_process import terminate_process_tree
+
+    _print(f"[cleanup] terminating {len(running)} running test(s) — {reason}")
+    for w_id, run_info in list(running.items()):
+        # Snapshot + kill the whole descendant tree. terminate_process_tree
+        # enumerates recursive children BEFORE killing the child pytest, so it
+        # still catches ManagedProcess workers that setsid into their own
+        # session — a process-group kill alone would miss them once the child
+        # pytest dies and they reparent to init.
+        try:
+            terminate_process_tree(run_info.proc.pid, immediate_kill=True, timeout=5)
+        except Exception as exc:  # never let one test's cleanup abort the rest
```

Ruff (0.15.21)

[warning] Do not catch blind exception: `Exception`

(BLE001)